docs/config/cluster.rst

```diff
@@ -1276,16 +1275,15 @@ The following settings can be used to configure the behavior of the
   | *Default:*   ``-1``
   | *Runtime:*  ``no``
 
-  The setting ``gateway.expected_data_nodes`` defines the number of data nodes
-  that should be waited for until the cluster state is recovered immediately.
-  The value of the setting should be equal to the number of data nodes in the
-  cluster, because you only want the cluster state to be recovered after all
-  data nodes are started.
+  The setting ``gateway.expected_data_nodes`` defines the total number of
+  data nodes expected in the cluster. It is evaluated together with 
+  ``gateway.recover_after_data_nodes``
+  to decide if the cluster can proceed with recovery.
 
 .. _gateway.recover_after_time:
 
 **gateway.recover_after_time**
-  | *Default:*   ``0ms``
+  | *Default:*   ``5m`` if ``expected_nodes`` or ``expected_data_nodes`` are set,  ``0ms`` otherwise
```

Comment:
This defaults to `5m`. There can't be 2 default values
 it is just not relevant  if no  expected nodes are set.

Reply:
No, it is like hernan described, the default value has a dependency.

Reply:
I think it is more complex then that ... Also the "default" would be `null` and not `0ms` , though being a technicality, it behaves slightly differently, then setting `0ms` explicitly.

```java
public static final TimeValue DEFAULT_RECOVER_AFTER_TIME_IF_EXPECTED_NODES_IS_SET = TimeValue.timeValueMinutes(5);
```

```java
if (RECOVER_AFTER_TIME_SETTING.exists(settings)) {
// explicitly set recover_after_time
    recoverAfterTime = RECOVER_AFTER_TIME_SETTING.get(settings);
} else if (expectedNodes >= 0 || expectedDataNodes >= 0 || expectedMasterNodes >= 0) {
// explicitly set expected nodes
    recoverAfterTime = DEFAULT_RECOVER_AFTER_TIME_IF_EXPECTED_NODES_IS_SET;
} else {
// default
    recoverAfterTime = null;
}
```

check later for `null`
```java
if (enforceRecoverAfterTime && recoverAfterTime != null) {
```

The setting is only relevant if
- set explicitly
_and/or_
- one of the expected nodes settings >= 1

How it behaves, is a combination of all 3 settings.
- `recover_after_nodes` are **always** a lower limit
- `expected_nodes` are an upper limit (unless `-1`) 
then: 
- if `expected_nodes` are set to `-1` (implicitly / explicitly)
  - if `recover_after_time` is not set implicitly => `null` -> immediate recover
  - if `recover_after_time` is set explicitly => recover after set time
- if `expected_nodes` are set to `0` or `1` 
   ->  immediate recovery independent (!) of setting
- if `expected_nodes` are set to  `>=1` 
  - if `recover_after_time` is not set implicitly => `5m` -> recover after `5m` if any node joins
  - if `recover_after_time` is set explicitly => recover after set time

---

```yaml
# recover_after_data_nodes: -1
# expected_data_nodes: -1
# recover_after_time: 5m
# --> recoverAfterTime == null
# --> recovers immediatily after any number of data nodes joined

recover_after_data_nodes: 2
# expected_data_nodes: -1
# recover_after_time: 5m
# --> recoverAfterTime == null
# --> recovers immediatily after 2 data nodes joined

# recover_after_data_nodes: -1
expected_data_nodes: -1
# recover_after_time: 5m
# --> recoverAfterTime == null
# --> recovers immediatily after any number of data nodes joined

# recover_after_data_nodes: -1
expected_data_nodes: 0
# recover_after_time: 5m
# --> recoverAfterTime == 5m
# --> recovers immediatily after any number of data nodes joined

# recover_after_data_nodes: -1
expected_data_nodes: 3
# recover_after_time: 5m
# --> recoverAfterTime == 5m
# --> recovers 5 minutes after any number of nodes joined
# --> recovers immediatily after 3 data nodes joined

# recover_after_data_nodes: -1
# expected_data_nodes: -1
recover_after_time: 3m
# --> recoverAfterTime == 3m
# --> recovers 3 minutes after any number of nodes joined

recover_after_data_nodes: 2
expected_data_nodes: 3
# recover_after_time: 5m
# --> recoverAfterTime == 5m
# --> recovers 5 minutes after 2 data nodes joined
# --> recovers immediatily after 3 data nodes joined

recover_after_data_nodes: 2
# expected_data_nodes: -1
recover_after_time: 3m
# --> recoverAfterTime == 3m
# --> recovers 3 minutes after 2 data nodes joined

# recover_after_data_nodes: 2
expected_data_nodes: 3
recover_after_time: 3m
# --> recoverAfterTime == 3m
# --> recovers 3 minutes after any number of nodes joined
# --> recovers immediatily after 3 data nodes joined

recover_after_data_nodes: 2
expected_data_nodes: 3
recover_after_time: 3m
# --> recoverAfterTime == 3m
# --> recovers 3 minutes after 2 data nodes joined
# --> recovers immediatily after 3 data nodes joined
```